Introduce basic .cursorrules for this repo

AliSoftware opened the pull request:
Since many of us are using Cursor IDE nowadays, I figured I'd share my initial `.cursorrules` files I created while working on this repo.

This is a very early draft / version of the rules I asked Cursor itself to come up with after analyzing our codebase of this repo + giving it some additional suggestions for rules about documentation and bash 3.2.57. But I'd love for you to add more ideas to that file so that we can consolidate those rules with our conventions and guidances we want for this repo.

---

- [ ] ~I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary~.

AliSoftware committed:
Introduce basic .cursorrules for this repo

iangmaia commented on .cursorrules:
````diff
@@ -0,0 +1,65 @@
+#   [Description of exit codes and their meanings]
+```
+
+## Bash 3.2 Compatibility
````
Perhaps it is worth it to explicitly mention a cross platform requirement, as very often I've seen mistakes on that front in generated code.

AliSoftware replied:
Sure, though I'd expect mentioning compatibility with Bash 3.2 would be enough to make it cross-platform? 🤔

iangmaia replied:
Though still it might bring in Linux-only / macOS-only **external tools** even if you mention Bash 3.2, no?

AliSoftware replied:
Ah, so you're talking about _external commands_ we might _invoke_ from those scripts, not bash builtins/functions/constructs/syntax, I see. Yes, good point, would be worth adding indeed.

There's already a "Document cross-platform considerations" item (L16) and "Document all dependencies and requirements" (L65), but we could definitively be more precise about that point, possibly with a dedicated `##` rule section for that in particular 👍

iangmaia commented on .cursorrules:
```diff
@@ -0,0 +1,65 @@
+## Bash 3.2 Compatibility
+Avoid using:
+- Associative arrays (declare -A)
+- mapfile/readarray command
```
```
- mapfile/readarray command
```
Very often it resorts to that, I wonder why 😅 
I very often write in prompts to avoid using external tools unless really necessary.

Maybe it makes sense, on top of an explicit "avoid list" of commands, add a "allow list" of common tools that are fine to use?

AliSoftware replied:
> I very often write in prompts to avoid using external tools unless really necessary.

`mapfile` and `readarray` are not _external tools_ (nor _commands_) though. They are bash built-ins. It's just that those built-ins were only introduced in bash 4.0+. Just like support for e.g. `declare -A` for associative arrays was only added in bash 4.0+ too.

And I'm not surprised that it often resorts to those 2 `mapfile` and `readarray` built-ins, because they are indeed really useful to read data into arrays in bash 4.0, while in bash 3.x we had to resort to more convoluted manual constructs like `while read` loops to achieve what `mapfile` and `readarray` can now achieve more concisely… as long as you run bash 4.0+ and not an earlier version like `3.2` that doesn't have those build-ins.

> Maybe it makes sense, on top of an explicit "avoid list" of commands, add a "allow list" of common commands that are fine to use?

Are you suggesting commands or bash build-ins there?

I don't see a point of us listing all the allowed bash 3.2 build-ins manually, as it'd be a very very long list, and I'd expect just mentioning bash 3.2 would be enough for Cursor to know what builtins are supported in that version of `bash`.

What I could see useful though, is indeed a list of _external commands_ we could allow it to use, like `jq` and the like, that are not built-ins but indeed external tools we have to install for them to be available, but we know are installed on most of our agents that will use those scripts.

iangmaia replied:
> What I could see useful though, is indeed a list of external commands we could allow it to use, like jq and the like

🎯 Yep, that's what I meant.

AliSoftware replied:
@iangmaia I've pushed a new commit to add additional rules about cross-platform, as well as a validation section, and asked Cursor to clean up the formatting. It should be more complete now 🙂

AliSoftware committed:
Add details about cross-platform considerations

And which commands we can assume are available in our agents for each type of script

AliSoftware committed:
Add Script Validation section

AliSoftware committed:
Improve .cursorrules documentation formatting and clarity

- Add backticks to code-like elements for better readability
- Standardize formatting of command and code references
- Enhance section formatting for improved visual consistency

AliSoftware committed:
Add rule about shebang

AliSoftware committed:
Reorder sections for consistency

iangmaia reviewed (approved):
Thank you for this 👍 it looks like a great starting point!
